calico/felix/fiptables.py

```diff
@@ -1,214 +1,451 @@
-    """
-    found = False
-    rules = chain.rules
-
-    if position == RULE_POSN_LAST:
-        position = len(rules)
-
-    if force_position:
-        if (len(rules) <= position) or (rule._rule != chain.rules[position]):
-            # Either adding after existing rules, or replacing an existing rule.
-            chain.insert_rule(rule._rule, position)
-    else:
-        #*********************************************************************#
-        #* The python-iptables code to compare rules does a comparison on    *#
-        #* all the relevant rule parameters (target, match, etc.) excluding  *#
-        #* the offset into the chain. Hence the test below finds whether     *#
-        #* there is a rule with the same parameters anywhere in the chain.   *#
-        #*********************************************************************#
-        if rule._rule not in chain.rules:
-            chain.insert_rule(rule._rule, position)
-
-    return
-
-
-def get_table(type, name):
-    """
-    Gets a table. This is a simple helper method that returns either
-    an IP v4 or an IP v6 table according to type.
-    """
-    if type == IPV4:
-        table = iptc.Table(name)
-    else:
-        table = iptc.Table6(name)
-
-    return table
-
-
-def get_chain(table, name):
-    """
-    Gets a chain, creating it first if it does not exist.
-    """
-    if table.is_chain(name):
-        chain = iptc.Chain(table, name)
-    else:
-        table.create_chain(name)
-        chain = iptc.Chain(table, name)
-
-    return chain
-
+# -*- coding: utf-8 -*-
+# Copyright 2014 Metaswitch Networks
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+felix.fiptables
+~~~~~~~~~~~~
+
+IP tables management functions. This is a wrapper round python-iptables that
+allows us to mock it out for testing.
+"""
+import logging
+import os
+import re
+import time
+
+from calico.felix import futils
+from calico.felix.futils import IPV4, IPV6
+
+# iptables command to use.
+IPTABLES_CMD = { IPV4: "iptables",
+                 IPV6: "ip6tables" }
+
+# Logger
+log = logging.getLogger(__name__)
+
+# Special value to mean "put this rule at the end".
+RULE_POSN_LAST = -1
+
+class Chain(object):
+    """
+    Representation of an iptables chain.
+    """
+    def __init__(self, table, name):
+        self.type = table.type
+        self.table = table
+        self.name = name
+        self.rules = []
+        self.table.chains[name] = self
```

Comment:
Also looks like there's a reference cycle here, which is best avoided in Python since it negates Python's reference counting and relies on Python's (relatively weak, stop-the-world) garbage collector.  Should Chain hold a weak ref to Table?